### PR TITLE
Axis now always show the min and max values. Also implemented snapping to valid values.

### PR DESCRIPTION
**Pull request by GeertvanHorrik**

### CODE REVIEW

**Comment by GeertvanHorrik** [comment]
Why the "in progress" label?

**Comment by tibel** [comment]
Adding a `Timer` to the axis?

**Comment by GeertvanHorrik** [comment]
Yes, the reason is that we need to snap after a while (when the user is scrolling, it should continue scrolling). It should then snap after x ms not to bother the user.

**Comment by GeertvanHorrik** [comment]
Otherwise the snapping will be too aggressive.

**Comment by tibel** [comment]
I thought that `PlotController` is for handling user interaction.

**Comment by GeertvanHorrik** [comment]
But it should always snap (even when set programmatically / etc). I am not really familiar with the controllers, so I might need some help. If you think there is a better solution, let me know, but I used a system timer to prevent any UI elements inside the axis.

**Comment by GeertvanHorrik** [comment]
Any updates on this?

**Comment by tibel** [comment]
Build server is working again.
With the change of the PCL core library to profile 259 you have no access to `Timer` class (see
https://forums.xamarin.com/discussion/17227/timer-in-portable-class-library).

**Comment by GeertvanHorrik** [comment]
Dear lord, no timer... Ok, will need a different way to solve this issue.

**Comment by GeertvanHorrik** [comment]
Now I remember, I implemented a timer in Catel for PCL:

https://github.com/Catel/Catel/blob/e8c26d4a8a7d09e2799f24f7f547650c4bf79433/src/Catel.Core/Catel.Core.Shared/Threading/Timer.cs

Can I put it in?

**Comment by tibel** [comment]
Yes, we have to include some internal Timer for PCL.

I still think we should not add a Timer to the axis and  would prefer one Timer per PlotModel or PlotController to handle the snapping. (Just my thoughts, no concrete idea how to implement this)

**Comment by GeertvanHorrik** [comment]
Let me start by adding a timer, then implement the way it currently is. Then we can always move it, right?

**Comment by tibel** [comment]
Sure

**Comment by GeertvanHorrik** [comment]
How can I inform a PlotModel that it must be updated from within an axis? We have the AxisChanged / TransformChanged events, but they are never subscribed to?

**Comment by GeertvanHorrik** [comment]
Extracting the snapping logic so it's not implemented inside Axis (but as an extension method).

**Comment by GeertvanHorrik** [comment]
Trying to move this to the controller. Requires additional events / methods to controller, but will keep it as a non-breaking change (feature additions only).

**Comment by GeertvanHorrik** [comment]
I have refactored the controller, but it's too complex, the controller doesn't have all events (for example, Pan, Zoom, etc). I think we really need to add this to the PlotModel instead.

**Comment by tibel** [comment]
I think we should ask @objorke for his input.

**Comment by GeertvanHorrik** [comment]
Now the collection changed stuff is added, I will finish this one.

**Comment by objorke** [comment]
I think the "in progress" tag was intended for "review in progress". First time I used a tag on a PR :-)

**Comment by GeertvanHorrik** [comment]
Ah. I thought it meant in progress by developer. They are set automatically, right?

**Comment by objorke** [comment]
Can the snapping feature be moved to a separate PR?

**Comment by objorke** [comment]
Please remove #region directives. http://docs.oxyplot.org/en/latest/contribute/coding-style.html

**Comment by objorke** [comment]
Can we have a flag to control whether the min/max values can be shown or not?

**Comment by GeertvanHorrik** [comment]
Not really. The reason for that is that for the min / max to be meaningful, we must snap. Otherwise you will get extremely weird min / max 3.0000000000045349 etc.

**Comment by objorke** [comment]
Please update the change log :-)

**Comment by GeertvanHorrik** [comment]
I think I was working on SnapSettings. I wanted to put them on Axis (so you can determine per axis how and if to snap and whether to show min / max values).

**Comment by objorke** [comment]
I see, I will test this later!
I am also a bit concerned about the timer being introduced here... Is it really needed?

**Comment by GeertvanHorrik** [comment]
Yes, we need a timer, unfortunately. I can get away with a Task.Delay, but that is uglier than a timer.

**Comment by GeertvanHorrik** [comment]
Sorry this is going to be so big :-(

**Comment by GeertvanHorrik** [comment]
I think it's ready for review.

**Comment by GeertvanHorrik** [comment]
I don't understand why it's failing, @objorke , can you take a look?

**Comment by objorke** [comment]
There are visual diffs after these changes:

Expected (baseline):
pngexportertests_exportwithresolution_192dpi

Actual:
pngexportertests_exportwithresolution_192dpi

Can this feature be implemented without doing a breaking change? Some users may want the current behavior (with 'padding' on the x-axis).

**Comment by GeertvanHorrik** [comment]
I thought I turned off the feature by default. Checking out why it's not. Will definitely keep it the way it was, so you can enable snapping if you want.

**Comment by GeertvanHorrik** [comment]
I couldn't get my unit tests to succeed. I undoed all my changes 1 by 1 to check what might be causing it, but none of them made the unit tests succeed. So decided to do a fresh branch from develop, but these unit tests also fail?

**Comment by GeertvanHorrik** [comment]
Found the issue, it was the refactoring of UpdateActualMinMax, where the calculation for the actual minimum requires the ActualMaximum to be set first.

**Comment by MerlinDuChaos** [comment]
When will this be available in the main branch ?  I'd need it for my project, this is a great addition ! +1

**Comment by objorke** [comment]
I can review this next week!

**Comment by GeertvanHorrik** [comment]
Awesome, thanks. Let me know if there is anything I can help you with in the meantime.

**Comment by GeertvanHorrik** [comment]
@objorke Hi Oystein, did you find time to review this?

**Comment by objorke** [comment]
Sorry for the delay, will review soon!

**Comment by objorke** [comment]
will try to find time in the weekend!

File: Source/OxyPlot/Graphics/Snapping.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+//   Snapping data.
+// </summary>
+// --------------------------------------------------------------------------------------------------------------------
+
+namespace OxyPlot
+{
+    /// <summary>
+    /// Snapping data.
+    /// </summary>
+    public class Snapping
```
Comment (objorke) [inline Source/OxyPlot/Graphics/Snapping.cs]: would it be better to add simple properties `IsSnappingEnabled` and `SnappingPrecision` to the axis? I am concerned about these snapping options being a mutable reference type. Also, I don't think it works well with the WPF axis "wrappers" - no change notifications if you change an `IsEnabled` property

**Comment by objorke** [comment]
I think there should be better definition for what snapping is doing.

File: Source/OxyPlot/PlotModel/PlotModel.cs [hunk 2]
```diff
@@ -193,6 +193,16 @@ public enum TitleHorizontalAlignment
...
     {
         /// <summary>
+        /// The snap timer.
+        /// </summary>
+        private readonly Timer snapTimer;
+
+        /// <summary>
+        /// Gets or sets whether the axis is currently snapping.
+        /// </summary>
+        private bool isSnapping;
```
Comment (objorke) [inline Source/OxyPlot/PlotModel/PlotModel.cs]: is this needed? should rather check if any axis has enabled snapping

File: Source/OxyPlot/PlotModel/PlotModel.cs [hunk 3]
```diff
@@ -193,6 +193,16 @@ public enum TitleHorizontalAlignment
     public partial class PlotModel : Model, IPlotModel
     {
         /// <summary>
+        /// The snap timer.
+        /// </summary>
+        private readonly Timer snapTimer;
```
Comment (objorke) [inline Source/OxyPlot/PlotModel/PlotModel.cs]: I don't like adding a timer. I think we discussed this before, must test. How will this work if you are printing the `PlotModel`? Should you wait a few timer intervals before printing? :-)

File: Source/OxyPlot/Utilities/Timer.cs [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    /// </summary>
+    /// <param name="state">
+    /// The state.
+    /// </param>
+    public delegate void TimerCallback(object state);
+
+    /// <summary>
+    /// Timer for WinRT since WinRT only provides the DispatcherTimer which cannot be used outside the UI thread.
+    /// </summary>
+    public class Timer : IDisposable
```
Comment (objorke) [inline Source/OxyPlot/Utilities/Timer.cs]: Cool, this is original code - not taken from anywhere, right?

Comment (objorke) [inline Source/OxyPlot/Utilities/Timer.cs]: But, I still hope we don't need a timer in this library. In my opinion timers should only be used in the platform specific projects (in the plot views).

File: Source/OxyPlot/OxyPlot_SL5.csproj [hunk 5]
```diff
@@ -30,7 +30,7 @@
     <Optimize>false</Optimize>
     <OutputPath>Bin\Debug\SL5\</OutputPath>
     <IntermediateOutputPath>obj\Debug\SL5\</IntermediateOutputPath>
-    <DefineConstants>DEBUG;TRACE;SILVERLIGHT</DefineConstants>
+    <DefineConstants>TRACE;DEBUG;SILVERLIGHT;SL5</DefineConstants>
```
Comment (objorke) [inline Source/OxyPlot/OxyPlot_SL5.csproj]: I think we can remove support for SL4. Will handle it in #115/#481

**Comment by objorke** [comment]
There should be on PR for the "show min/max values" issue #19. This one should be easy to approve.
There should be another PR for the "snapping" feature and a corresponding issue. 
@GeertvanHorrik Can you create the issue and split the PR?
The 29 commits must also be squashed. I would like to see 1 commit on each PR/branch!
Sorry that this review has been taking so long.

File: CHANGELOG.md [hunk 6]
```diff
@@ -15,6 +15,9 @@ All notable changes to this project will be documented in this file.
 - Added capability to display mean value to BoxPlot (#440)
 - Added TitleToolTip to PlotModel (#508)
 - Expose PlotElement's TextColor property on WPF.Axes.Axis (#452)
+- Added snapping feature to PlotModel and Axis. This also ensures that the minimum and maximum values 
+  on the axis are always displayed (#19, #519). This feature is disabled by default, but can be enabled
```
Comment (objorke) [inline CHANGELOG.md]: I think we should try to only reference issues, not PRs

File: Source/OxyPlot/Graphics/Snapping.cs (hunk 1 above)
Comment (GeertvanHorrik) [inline Source/OxyPlot/Graphics/Snapping.cs]: The reason I used an object is not to cause breaking changes in the future. If you want new snapping data in the future, you can add it without a breaking change or making a mess. But happy to make it direct properties (they were direct properties in the first place).

File: Source/OxyPlot/OxyPlot_SL5.csproj (hunk 5 above)
Comment (GeertvanHorrik) [inline Source/OxyPlot/OxyPlot_SL5.csproj]: But even then, WP80 / WP80SL also define the SILVERLIGHT definition so I still think SL5 is a good define constant.

File: Source/OxyPlot/PlotModel/PlotModel.cs (hunk 3 above)
Comment (GeertvanHorrik) [inline Source/OxyPlot/PlotModel/PlotModel.cs]: I know, I wish there was another (easy) way. But because the timer is inside the model, not inside the control, it will also work during printing. And... the timer will only kick in when you pan / zoom / navigate anyway.

File: Source/OxyPlot/PlotModel/PlotModel.cs (hunk 2 above)
Comment (GeertvanHorrik) [inline Source/OxyPlot/PlotModel/PlotModel.cs]: This is required. Snapping might be enabled, but the axis might not be snapping *at the moment*. We need to prevent that because of the snapping "move", we don't spawn another snapping call (and cause unnecessary calls).

File: Source/OxyPlot/Utilities/Timer.cs (hunk 4 above)
Comment (GeertvanHorrik) [inline Source/OxyPlot/Utilities/Timer.cs]: Not original code, it comes from Catel (I wrote it, have to fight all the platform differences there as well). It's MIT licensed so you can do whatever you want with it:

https://github.com/Catel/Catel/blob/develop/src/Catel.Core/Catel.Core.Shared/Threading/Timer.cs

**Comment by GeertvanHorrik** [comment]
First of all, thanks for reviewing. I know we are all busy so don't worry too much about it. I will squash everything, but about the 2 separate PR:

I have created the "show min / max values on the scale". That was the initial issue I was trying to solve. Then I saw a *lot* of nonsense values like 0.000000099, etc. Then I decided: I can round up these values and lie about them or I can implement a snapping (that's what users actually want IMHO). So I wanted to snap. But, unfortunately this needed a timer (because when you are panning, you want to snap *once* you are ready panning, etc).

The bottom line is that I am not sure if one can go without the other. How do you want to proceed with this?

**Comment by objorke** [comment]
Instead of a timer, can't there be a "panning complete" event?

**Comment by objorke** [comment]
Sorry for the delays and being very picky on this PR... There is something I don't like with this design, but I don't have a clear picture of how it should be yet. Maybe the snapping functionality should be the responsibility of the Views, the `PlotController` or the `PanManipulator`. I need to look into the code...
Any others with opinions on this implementation? @tibel?

**Comment by GeertvanHorrik** [comment]
_Np, I am happy that you are picky (we all need to be, we want a good product, right :-) )_

Anyway, I tried implementing this inside the controller as well, but when is panning assumed completed? When I hold down the left arrow, when I hit the left arrow 7 times fast in burst mode, etc.

The only thing I can think of is to move the timer to the controller instead of the axis. Then if someone modifies the plot manually, they need to call .Snap() manually, which is fine by me.

**Comment by objorke** [comment]
Right, it is difficult to know when you are done panning with the keyboard... :-) But isn't it ok to snap after each keystroke? The keyboard panning could use the tick interval lengths as "delta values"!

Still no example included in this PR? I created the following model in the example library:
```csharp
        [Example("Axis snapping")]
        public static PlotModel AxisSnapping()
        {
            var plotModel1 = new PlotModel { Title = "Axis snapping" };
            var axis1 = new LinearAxis { Position = AxisPosition.Left };
            var axis2 = new LinearAxis { Position = AxisPosition.Bottom };
            axis1.Snapping.IsEnabled = true;
            axis2.Snapping.IsEnabled = true;
            plotModel1.Axes.Add(axis1);
            plotModel1.Axes.Add(axis2);
            return plotModel1;
        }
```

Regarding the Snapping class, I think it would be nicer if I could write
```csharp
plotModel1.Axes.Add(new LinearAxis { Position = AxisPosition.Left, IsSnappingEnabled = true });
```

I notice that after panning, the axes show numbers like 46, 66, 86 when snapping is enabled, but nice numbers like 40, 60, 80 when snapping is disabled. Was this intended?

What are good values for `Precision` (not auto) for the example above?

Is it possible that we can support this feature by adding some extension points in the plot controller? Virtual methods or events? I am still not convinced that it is a good idea to add this feature in the general library, and particularly not on the model.

**Comment by GeertvanHorrik** [comment]
So to summarize:

1. You want the snapping class to be removed and just be a property directly on the axis?
2. Precision is automatically calculated based on the values the axis is displaying in the current zoom factor. I made it overridable in case it was not behaving as expected so it can always be "hotfixed" by the end-users

What if I remove all the timing stuff and just add the Snap function. Then it's the responsibility of the end-developer to implement this. I will implement it with a timer on my side.

**Comment by objorke** [comment]
I am concerned about the maintainability of this feature in the core library, it is making the "update" pass more complicated.
My preference would be to add some virtual methods or events in the core, then include an example showing how to do snapping. It would be great if such an example was created in the portable example library.
Yes, in my opinion the timing stuff should be removed. Snapping should happen immediately when panning (also by keys) has completed.

**Comment by GeertvanHorrik** [comment]
That's why I don't want anything inside the update pass. I just want a Snap() method on the axis so we can manually call it, that's all. I think that can be put into the core?

**Comment by objorke** [comment]
I don't quite understand... Can we call it something more general than `Snap`?

**Comment by GeertvanHorrik** [comment]
Of course. Anything you want (I will probably put this into my custom usage of OxyPlot anyway). The only thing I needed in the first place is that the outer values on an axis are visible (by default, but also after dragging). Since you can have some nonsense value like 3.00000000000000e2 or something like that which make no sense to an end-user, I wanted to snap such values. That's why I called it snapping in the first place, but other names such as ResizeViewPortToSensibleValues or whatever you prefer.

**Comment by objorke** [comment]
Sorry, I am still not happy with the name - I wanted a general method or event where you could extend the custom functionality (like snapping or something else). But maybe I don't understand correctly... Where should this method or event live? Where should it be called from?

I looked at #19 again, I think snapping is not part of that issue. krafty wanted to see the minimum value of the data, not a 'nice' number.

I see you comment from Aug 20, but I still think #19 and snapping are different features. What about creating a new issue about snapping with clear requirements about what it should do? I understand it as
- snapping: adjust the actual minimum and maximum of an axis to values at major ticks (at 'nice' numbers)
- do the snapping (adjustment) when minimum and maximum has not been defined 
- do the snapping after the axis has been panned or zoomed

I hope we can find a good solution soon!

**Comment by GeertvanHorrik** [comment]
Yes, you want to show the display values of the min and max. But does this number make sense?

3.9e23 or would the end user (and #19) want to see 4 in that case? That's why I introduced snapping. Snapping to logical values that make sense for the values that are actually representing the axis.

If you can come up with a nice name, I am happy to use that as well. Maybe TuneAxis() which is virtual? By default it doesn't do anything, but you can override to implement features like snap.

**Comment by GeertvanHorrik** [comment]
So..... the verdict is? I am happy to close this one. I will implement it in my own solution instead, I Think it's too complex to be implemented in the core.

**Comment by objorke** [comment]
Let us close this PR if it is OK for you, I am not comfortable merging this code yet. [Jeremy Miller](http://jeremydmiller.com/) has a good post about such issues, see the section [Taking pull requests](http://jeremydmiller.com/2015/10/15/thoughts-on-running-an-oss-project/).

I am willing to add "extension points" where you can implement functionality like snapping, but let us create a separate issue/PR for that. I think the important question is *where* the extension point (virtual method, event) should be called from.

I have started looking into #19, will continue discussion on min/max values there (I think 3.9e23 is the correct value to show!)
